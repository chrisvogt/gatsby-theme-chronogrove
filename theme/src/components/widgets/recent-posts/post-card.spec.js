import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen, fireEvent } from '@testing-library/react'
import PostCard from './post-card'

// Mock window.scrollTo
const mockScrollTo = jest.fn()
Object.defineProperty(window, 'scrollTo', {
  value: mockScrollTo,
  writable: true
})

// Mock Gatsby navigation
const mockNavigate = jest.fn()
Object.defineProperty(window, '___navigate', {
  value: mockNavigate,
  writable: true
})

describe('PostCard', () => {
  const baseProps = {
    banner: 'https://cdn.example.com/images/article-og-banner.jpg',
    category: 'personal',
    date: '1592202624',
    link: '/blog/article',
    title: 'My Blog Post'
  }

  it('matches the snapshot without excerpt', () => {
    const tree = renderer.create(<PostCard {...baseProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders excerpt when provided', () => {
    const propsWithExcerpt = {
      ...baseProps,
      excerpt: 'This is a sample excerpt for the blog post.'
    }
    const tree = renderer.create(<PostCard {...propsWithExcerpt} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('does not render excerpt when explicitly null', () => {
    const propsWithNullExcerpt = {
      ...baseProps,
      excerpt: null
    }
    const tree = renderer.create(<PostCard {...propsWithNullExcerpt} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('does not render excerpt when explicitly undefined', () => {
    const propsWithUndefinedExcerpt = {
      ...baseProps,
      excerpt: undefined
    }
    const tree = renderer.create(<PostCard {...propsWithUndefinedExcerpt} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without banner when not provided', () => {
    const propsWithoutBanner = {
      category: 'personal',
      date: '1592202624',
      link: '/blog/article',
      title: 'My Blog Post'
    }
    const tree = renderer.create(<PostCard {...propsWithoutBanner} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without category when not provided', () => {
    const propsWithoutCategory = {
      banner: 'https://cdn.example.com/images/article-og-banner.jpg',
      date: '1592202624',
      link: '/blog/article',
      title: 'My Blog Post'
    }
    const tree = renderer.create(<PostCard {...propsWithoutCategory} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('specifically tests excerpt prop handling for blog page change', () => {
    // This test specifically addresses the user's change where excerpt was commented out
    const propsWithoutExcerpt = {
      ...baseProps
      // excerpt is intentionally not included, simulating the commented out line
    }

    const component = renderer.create(<PostCard {...propsWithoutExcerpt} />)
    const tree = component.toJSON()

    // Verify that no excerpt paragraph is rendered
    const excerptParagraph = findExcerptParagraph(tree)
    expect(excerptParagraph).toBeNull()

    // Verify that the component still renders correctly without excerpt
    expect(tree).toBeTruthy()
  })

  it('renders in horizontal layout when horizontal prop is true', () => {
    const propsWithHorizontal = {
      ...baseProps,
      horizontal: true
    }
    const tree = renderer.create(<PostCard {...propsWithHorizontal} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders as image-only recap when isRecap prop is true', () => {
    const propsWithRecap = {
      ...baseProps,
      isRecap: true
    }
    const tree = renderer.create(<PostCard {...propsWithRecap} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls window.scrollTo when clicked', () => {
    mockScrollTo.mockClear()
    mockNavigate.mockClear()

    render(<PostCard {...baseProps} />)
    const link = screen.getByRole('link')

    fireEvent.click(link)

    expect(mockScrollTo).toHaveBeenCalledWith(0, 0)
  })
})

// Helper function to find the excerpt paragraph in the rendered tree
function findExcerptParagraph(tree) {
  if (!tree || !tree.children) return null

  for (const child of tree.children) {
    if (child.type === 'p' && child.props && child.props.className && child.props.className.includes('description')) {
      return child
    }
    if (child.children) {
      const found = findExcerptParagraph(child)
      if (found) return found
    }
  }
  return null
}
