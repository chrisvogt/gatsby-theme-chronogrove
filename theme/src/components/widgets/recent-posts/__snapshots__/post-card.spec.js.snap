// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PostCard does not render excerpt when explicitly null 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-xz12ce"
  >
    <div
      className="card-content css-ihz6y5"
    >
      <div
        className="card-media css-1usooy4"
      >
        <div
          className="css-1kkguim"
          role="img"
        />
      </div>
      <div
        className="css-iiqpqc"
      >
        <div
          className="css-ulp4wu"
        >
          Personal
        </div>
        <h3
          className="css-1t3cywp"
        >
          My Blog Post
        </h3>
        <time
          className="created css-1c6hy0j"
        >
          1592202624
        </time>
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard does not render excerpt when explicitly undefined 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-xz12ce"
  >
    <div
      className="card-content css-ihz6y5"
    >
      <div
        className="card-media css-1usooy4"
      >
        <div
          className="css-1kkguim"
          role="img"
        />
      </div>
      <div
        className="css-iiqpqc"
      >
        <div
          className="css-ulp4wu"
        >
          Personal
        </div>
        <h3
          className="css-1t3cywp"
        >
          My Blog Post
        </h3>
        <time
          className="created css-1c6hy0j"
        >
          1592202624
        </time>
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard matches the snapshot without excerpt 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-xz12ce"
  >
    <div
      className="card-content css-ihz6y5"
    >
      <div
        className="card-media css-1usooy4"
      >
        <div
          className="css-1kkguim"
          role="img"
        />
      </div>
      <div
        className="css-iiqpqc"
      >
        <div
          className="css-ulp4wu"
        >
          Personal
        </div>
        <h3
          className="css-1t3cywp"
        >
          My Blog Post
        </h3>
        <time
          className="created css-1c6hy0j"
        >
          1592202624
        </time>
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard renders as image-only recap when isRecap prop is true 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-xz12ce"
  >
    <div
      className="card-content css-ihz6y5"
    >
      <div
        className="card-media css-1liniix"
      >
        <div
          aria-label="My Blog Post"
          className="css-1kkguim"
          role="img"
          title="My Blog Post"
        />
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard renders as image-only recap with horizontal layout 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-nwtitb"
  >
    <div
      className="card-content css-1taqpx2"
    >
      <div
        className="card-media css-1liniix"
      >
        <div
          aria-label="My Blog Post"
          className="css-1qp8vmu"
          role="img"
          title="My Blog Post"
        />
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard renders excerpt when provided 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-xz12ce"
  >
    <div
      className="card-content css-ihz6y5"
    >
      <div
        className="card-media css-1usooy4"
      >
        <div
          className="css-1kkguim"
          role="img"
        />
      </div>
      <div
        className="css-iiqpqc"
      >
        <div
          className="css-ulp4wu"
        >
          Personal
        </div>
        <h3
          className="css-1t3cywp"
        >
          My Blog Post
        </h3>
        <time
          className="created css-1c6hy0j"
        >
          1592202624
        </time>
        <p
          className="description css-1uz48mf"
        >
          This is a sample excerpt for the blog post.
        </p>
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard renders in horizontal layout when horizontal prop is true 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-nwtitb"
  >
    <div
      className="card-content css-1taqpx2"
    >
      <div
        className="card-media css-1usooy4"
      >
        <div
          className="css-1qp8vmu"
          role="img"
        />
      </div>
      <div
        className="css-p08ytg"
      >
        <div
          className="css-ulp4wu"
        >
          Personal
        </div>
        <h3
          className="css-hso91r"
        >
          My Blog Post
        </h3>
        <time
          className="created css-1c6hy0j"
        >
          1592202624
        </time>
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard renders without banner when not provided 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-xz12ce"
  >
    <div
      className="card-content css-ihz6y5"
    >
      <div
        className="css-iiqpqc"
      >
        <div
          className="css-ulp4wu"
        >
          Personal
        </div>
        <h3
          className="css-1t3cywp"
        >
          My Blog Post
        </h3>
        <time
          className="created css-1c6hy0j"
        >
          1592202624
        </time>
      </div>
    </div>
  </div>
</a>
`;

exports[`PostCard renders without category when not provided 1`] = `
<a
  className="css-sod6qt"
  href="/blog/article"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  <div
    className="css-xz12ce"
  >
    <div
      className="card-content css-ihz6y5"
    >
      <div
        className="card-media css-1usooy4"
      >
        <div
          className="css-1kkguim"
          role="img"
        />
      </div>
      <div
        className="css-iiqpqc"
      >
        <h3
          className="css-1t3cywp"
        >
          My Blog Post
        </h3>
        <time
          className="created css-1c6hy0j"
        >
          1592202624
        </time>
      </div>
    </div>
  </div>
</a>
`;
